Updated Non Secure and Secure Readme. (#18)
FreeRTOS/lab-iot-reference-stm32u5: Projects/b_u585i_iot02a_ntz/Readme.md

# Getting Started Guide for STM32U5 IoT Discovery Kit with AWS

## Overview

> The following Readme.md contains instructions on getting the non-secure(b_u585i_iot02a_ntz) version of the project up and running. It connects to AWS Core and publishes data. It also contains instructions on how to perform an OTA update in the later section of the Readme.md. 

Reviewer: avoid using "non-secure" project

Hit *Next*. The following prompt will pop up:

<img width="550" alt="16" src="https://user-images.githubusercontent.com/44592967/153657341-fa9c526f-b25d-40cd-ae05-d02e8f3551b5.PNG">

Click on *Browse* above and navigate to the root of the project. Click on *Finish*.

![image](https://user-images.githubusercontent.com/44592967/162328169-c3820c47-a21f-4dac-85fb-e69cd8aa184c.png)

This is how the workspace with the imported project looks like:

![image](https://user-images.githubusercontent.com/44592967/162328220-3d5a6f65-589d-4054-a52b-0f6584f46d68.png)

Build the b_u585i_iot02a_ntz project and flash the binary by clicking on *Run->Run*. Make sure that the board is plugged in.

 ## Running the demo

 With the firmware flashed to the board, open a command prompt, and navigate to the root of the project(lab-iot-reference-stm32u5). Type:

 `python tools/provision.py -i -v `

To know more about the above command, visit the Troubleshooting section at the end of the document.

The script will prompt you to enter the following details. You only need to update wifi_ssid, wifi_credential, mqtt_endpoint and thing name.

```

> tls_verify_ca[]: <Click enter>

Reviewer 2: ```suggestion
```

time_hwm[]: <Click enter>

> Wifi_credential[]: <your wifi password>

Reviewer 2: ```suggestion
wifi_credential[]: <your wifi password>
```

wifi_auth[]: <Click enter>
mqtt_endpoint[]: <a1qwhobjt*****-ats.iot.us-east-2.amazonaws.com>
wifi_ssid[]: <your wifi ssid>

> tls_verify[]: <Click enter>

Reviewer 2: ```suggestion
```

### Provision Code Signing credentials

Devices uses digital signatures to verify the authenticity of the firmware updates sent over the air. Images are signed by an authorized source who creats the image, and device can verify the signature of the image, using the corresponding public key of the source. Steps below shows how to setup and provision the code signing credentials so as to enable cloud to digitally sign the image and the device to verify the image signature before boot.

1. In your working directory, use the following text to create a file named *cert_config.txt*. Replace *test_signer@amazon.com* with your email address:

```
[ req ]
prompt             = no
distinguished_name = my_dn

[ my_dn ]
commonName = test_signer@amazon.com

[ my_exts ]
keyUsage         = digitalSignature
extendedKeyUsage = codeSigning
```
2. Create an ECDSA code-signing private key:

```

> openssl genpkey -algorithm EC -pkeyopt ec_paramgen_curve:P-256 -pkeyopt ec_param_enc:named_curve -outform PEM -out ecdsasigner-priv-key.pem

Reviewer 2: ```suggestion
openssl genpkey -algorithm EC -pkeyopt ec_paramgen_curve:P-256 -pkeyopt ec_param_enc:named_curve -outform PEM -out ecdsasigner-priv-key.pem
openssl ec -inform pem -in ecdsasigner-priv-key.pem -pubout -outform pem -out ecdsasigner-pub-key.pem
```

```

> 3.Create an ECDSA code-signing certificate:

Reviewer 2: ```suggestion
3.Create an ECDSA code-signing certificate to be uploaded to AWS IoT Core:
```

```
openssl req -new -x509 -config cert_config.txt -extensions my_exts -nodes -days 365 -key ecdsasigner-priv-key.pem -out ecdsasigner.crt

```

4. Import the code-signing certificate and private key into AWS Certificate Manager:

NOTE: This command displays an ARN for your certificate. You need this ARN when you create an OTA update job later

```

> aws acm import-certificate --certificate fileb://ecdsasigner.crt --private-key fileb://ecdsasigner.key

Reviewer 2: ```suggestion
aws acm import-certificate --certificate fileb://ecdsasigner.crt --private-key fileb://ecdsasigner-priv-key.pem
```

```

5. Connect the device to a terminal over serial port. On the command line prompt type following command to provision public key to device:

>   `> pki import cert ota_signer_pub`

Reviewer 2: ```suggestion
  `> pki import key ota_signer_pub`
```

`ota_signer_pub` is the label used to refer to the code signing key during verification of the firmware update.

6. Create a signing profile in AWS to sign the firmware image.

```
aws signer put-signing-profile --profile-name <your profile name> --signing-material certificateArn=<certificate arn created in step 4> --platform AmazonFreeRTOS-Default --signing-parameters certname=ota_signer_pub
```

## Setup OTA S3 bucket, Service role and policies in AWS

1. S3 bucket is used to store the new firmware image to be updated. To create a new S3 bucket follow these steps here: https://docs.aws.amazon.com/freertos/latest/userguide/dg-ota-bucket.html

2. Create a service role which grants permission for OTA service to access the firmware image: https://docs.aws.amazon.com/freertos/latest/userguide/create-service-role.html

3. Create an OTA update policy using the documentatio here: https://docs.aws.amazon.com/freertos/latest/userguide/create-ota-user-policy.html

4. Add a policy for AWS IoT to access the code signing profile: https://docs.aws.amazon.com/freertos/latest/userguide/code-sign-policy.html

## Create a code signed firmware update job

> 1. Bump up the version of the new firmware image to be updated. From the demo project, open File B-U585I-IOT02A/tz_disabled/Inc/ota_config.h and set APP_VERSION_MAJOR to 1 higher than current version. Build the firmware image using STM32Cube IDE.

Reviewer 2: ```suggestion
1. Bump up the version of the new firmware image to be updated. From the demo project, open File Common/config/ota_config.h and set APP_VERSION_MAJOR to 1 higher than current version. Build the firmware image using STM32Cube IDE.
```

 ```
 aws iot create-ota-update --cli-input-json file://<ota job configuration file path in your filesystem>
 ```

 The command on success returns the OTA Update identifier and status of the Job as *CREATE_PENDING*. To get the corresponding job ID of the OTA Job, execute the following command and look for *awsIotJobId* field in json document returned.

 ```
 aws iot get-ota-update --ota-update-id=<ota update id created above>
 ```
 Note down the job ID to check the status of the job later.

 ## Monitoring and Verification of firmware update

 1. Once the job is created on the terminal logs, you will see that OTA job is accepted and device starts downloading image.

```
>>>>>>> cfc299a (Updating Getting Started guide)
> AF>    16351 [OTAAgent] Current State=[WaitingForFileBlock], Event=[RequestFileBlock], New state=[WaitingForFileBlock] (ota.c:2834)
> AF>    15293 [OTAAgent] Extracted parameter: [key: value]=[execution.jobDocument.afr_ota.streamname: AFR_OTA-eb53bc47-6918-4b2c-9c85-a4c74c44a04c] (ota.c:1642)
<INF>    15294 [OTAAgent] Extracted parameter: [key: value]=[execution.jobDocument.afr_ota.protocols: ["MQTT"]] (ota.c:1642)

> <INF>    15296 [OTAAgent] Extracted parameter: [key: value]=[filepath: tz_disabled.bin] (ota.c:1642)

Reviewer 2: ```suggestion
<INF>    15296 [OTAAgent] Extracted parameter: [key: value]=[filepath: b_u585i_iot02a_ntz.bin] (ota.c:1642)
```
